### PR TITLE
[CWS][SEC-7286] Security Profiles manager with local watcher

### PR DESCRIPTION
**Pull request by Gui774ume**
### What does this PR do?

This PR introduces the security profile manager. This manager has 2 key functions:
- parse Security Profiles from the local file system
- match the monitored workloads with their relevant Security Profiles

This PR also introduces the ability to generate Security Profiles from a dump. 2 changes were made in the encoding of activity dumps:
- `Profile` was renamed to `SecL`
- `Profile` no indicates that an activity dump should be encoded into the Security Profile protobuf definition

New configuration parameters were added:
- `runtime_security_config.security_profile.enabled` : defines if the security profile manager should be enabled
- `runtime_security_config.security_profile.dir` : defines the directory in which the security profiles are stored (defaults to `/etc/datadog-agent/runtime-security.d/profiles`)
- `runtime_security_config.security_profile.watch_dir` : defines if the runtime watcher should be enabled
- `runtime_security_config.security_profile.cache_size` : defines the user space cache size for security profiles

### Motivation

This PR is a building block for upcoming improvements in the way to use Security Profiles on the agent side.

### Describe how to test/QA your changes

- Generate a simple activity dump and specify the `Profile` encoding.
- Copy the generated `.profile` file in the default `/etc/datadog-agent/runtime-security.d/profiles/` directory.
- Make sure the agent picks up on it and successfully loaded the profile (there should be a log).
- Remove the profile and make sure the agent properly cleaned up the cache (there should be a log).
- Add multiple profiles of different versions and make sure we properly keep only the last one.
- Add files with different extensions and make sure they're not read.
- Make sure the permissions of both the security profiles and the security profiles folder are properly checked (the folder should be `0750` and each security profile should be `0400`).

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [x] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [x] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [x] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [x] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/security/agent/agent.go [hunk 1]
```diff
@@ -41,7 +41,7 @@ type RuntimeSecurityAgent struct {
 	cancel               context.CancelFunc
 
 	// activity dump
-	storage *activitydump.ActivityDumpStorageManager
+	storage *activitydump2.ActivityDumpStorageManager
```
Comment (YoannGh) [inline pkg/security/agent/agent.go:44]: can't you use the `dump` package name here?

Comment (Gui774ume) [reply pkg/security/agent/agent.go:44]: Rrrrrr ... my IDE automatically does this, updating ! Thanks !

File: pkg/security/tests/activity_dumps_test.go [hunk 2]
```diff
@@ -85,7 +86,7 @@ func TestActivityDumps(t *testing.T) {
 			t.Fatal(err)
 		}
 
-		validateActivityDumpOutputs(t, test, expectedFormats, dump.OutputFiles, func(ad *activitydump.ActivityDump) bool {
+		validateActivityDumpOutputs(t, test, expectedFormats, dump.OutputFiles, func(ad *activitydump2.ActivityDump) bool {
```
Comment (YoannGh) [inline pkg/security/tests/activity_dumps_test.go:89]: same thing here

Comment (Gui774ume) [reply pkg/security/tests/activity_dumps_test.go]: 🤦🏻

File: pkg/security/proto/security_profile/v1/activity_dump.proto [hunk 3]
```diff
@@ -33,6 +33,15 @@ message Metadata {
     string serialization = 14;
 }
 
+message SecurityProfile {
+    string status = 1;
+    string version = 2;
+    Metadata metadata = 3;
```
Comment (spikat) [inline pkg/security/proto/security_profile/v1/activity_dump.proto:39]: Should we split/reduce the used Metadata to avoid having useless fields sent back as profile ? (agent version/commit, kernel, linux, arch, start, end, size ..)

File: pkg/security/security_profile/dump/activity_dump.go [hunk 4]
```diff
@@ -8,7 +8,7 @@
 
 //go:generate go run github.com/mailru/easyjson/easyjson -gen_build_flags=-mod=mod -no_std_marshalers -build_tags linux $GOFILE
 
-package activitydump
+package dump
```
Comment (paulcacheux) [inline pkg/security/security_profile/dump/activity_dump.go:11]: All those file moves will create a lot of churn and conflict, can't we do that in a separate PR ?

**Review by sgnn7 — APPROVED** [review]
LGTM for ASC-owned `pkg/config/config.go`

File: pkg/security/security_profile/profile/profile_dir.go [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+	"golang.org/x/exp/slices"
+
+	proto "github.com/DataDog/datadog-agent/pkg/security/proto/security_profile/v1"
+	cgroupModel "github.com/DataDog/datadog-agent/pkg/security/resolvers/cgroup/model"
+	"github.com/DataDog/datadog-agent/pkg/security/seclog"
+	"github.com/DataDog/datadog-agent/pkg/security/utils"
+)
+
+var (
+	debounceDelay        = 5 * time.Second
```
Comment (spikat) [inline pkg/security/security_profile/profile/profile_dir.go:32]: will be more clear if called workloadSelectorDebounceDelay ?

File: pkg/security/security_profile/profile/profile_dir.go [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+}
+
+// Start runs the directory provider
+func (dp *DirectoryProvider) Start(ctx context.Context) error {
+	dp.workloadSelectorDebouncer.Start()
+	dp.newFilesDebouncer.Start()
+
+	// start by loading the profiles in the configured directory
+	if err := dp.loadProfiles(); err != nil {
+		return fmt.Errorf("couldn't scan the security profiles directory: %w", err)
```
Comment (spikat) [inline pkg/security/security_profile/profile/profile_dir.go:107]: Is there a race if we receive a profile after loadProfiles() but before adding the fsnotify watcher?
Unlikely, but we may want to swap these two to avoid it

File: pkg/security/security_profile/dump/activity_dump.go (hunk 4 above)
Comment (Gui774ume) [reply pkg/security/security_profile/dump/activity_dump.go:11]: I _**think**_ we're good: the "only" changes we're expecting to make in the coming weeks are related to Security Profiles and will require at least this PR (which I hope we can merge soon). TL;DR I'll make another PR if we get just one conflict ! Deal ?

File: pkg/security/security_profile/profile/profile_dir.go (hunk 5 above)
Comment (Gui774ume) [reply pkg/security/security_profile/profile/profile_dir.go:32]: 👍🏻 on it !

**Review by brycekahle — APPROVED** [review]
shallow review on ebpf-platform owned files only

File: pkg/config/system_probe.go [hunk 7]
```diff
@@ -294,9 +294,15 @@ func InitSystemProbeConfig(cfg Config) {
 	cfg.BindEnvAndSetDefault("runtime_security_config.activity_dump.max_dump_count_per_workload", 25)
 	cfg.BindEnvAndSetDefault("runtime_security_config.activity_dump.tag_rules.enabled", false)
 
-	// SBOM
+	// CWS - SBOM
 	cfg.BindEnvAndSetDefault("runtime_security_config.sbom.enabled", false)
 	cfg.BindEnvAndSetDefault("runtime_security_config.sbom.workloads_cache_size", 10)
+
+	// CWS - Security Profiles
+	cfg.BindEnvAndSetDefault("runtime_security_config.security_profile.enabled", false)
```
Comment (paulcacheux) [inline pkg/config/system_probe.go:302]: Not a huge fan of this, I think this will cause confusion because activity dumps are kind of already called security profiles in the frontend.

File: pkg/security/config/dump.go [hunk 8]
```diff
@@ -107,15 +107,17 @@ var (
...
+	// SecL is used to request the Secl policy format
+	SecL StorageFormat = "secl"
 
 	strToFormats = make(map[string]StorageFormat)
 )
 
 // AllStorageFormats returns the list of supported formats
 func AllStorageFormats() []StorageFormat {
-	return []StorageFormat{JSON, PROTOBUF, DOT, Profile}
+	return []StorageFormat{JSON, PROTOBUF, DOT, Profile, SecL}
```
Comment (paulcacheux) [inline pkg/security/config/dump.go:120]: Why do we have a mix of UPPER and Mixed case here ? Since you are touching this, could we sync on UPPER only ?

File: pkg/security/security_profile/dump/activity_dump.go [hunk 9]
```diff
@@ -279,6 +279,37 @@ func NewActivityDumpFromMessage(msg *api.ActivityDumpMessage) (*ActivityDump, er
...
+	}
+
+	for node != nil {
+		for _, nr := range node.Syscalls {
+			mask[nr] = 1
+		}
+		for _, child := range node.Children {
+			nodes = append(nodes, child)
+		}
+		nodes = append(nodes[1:])
```
Comment (paulcacheux) [inline pkg/security/security_profile/dump/activity_dump.go:298]: Not sure, but this seems wrong

File: pkg/security/security_profile/dump/activity_dump.go [hunk 10]
```diff
@@ -279,6 +279,37 @@ func NewActivityDumpFromMessage(msg *api.ActivityDumpMessage) (*ActivityDump, er
...
+		if len(nodes) > 0 {
+			node = nodes[0]
+		} else {
+			node = nil
+		}
+	}
+
+	var output []uint32
```
Comment (paulcacheux) [inline pkg/security/security_profile/dump/activity_dump.go:306]:
```suggestion
	output := make([]uint32, 0, len(masks))
```

File: pkg/security/security_profile/dump/manager_test.go [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+func compareListOfDumps(t *testing.T, out, expectedOut []*ActivityDump) {
+	for _, elem := range out {
+		var found bool
+		for _, expected := range expectedOut {
+			if elem.Name == expected.Name {
+				found = true
+			}
+		}
+
+		assert.Equalf(t, true, found, "output didn't match: dump %s should not be in the output", elem.Name)
```
Comment (paulcacheux) [inline pkg/security/security_profile/dump/manager_test.go:32]:
```suggestion
		assert.Truef(t, found, "output didn't match: dump %s should not be in the output", elem.Name)
```

File: pkg/security/security_profile/dump/manager_test.go [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+	}
+	for _, elem := range expectedOut {
+		var found bool
+		for _, got := range out {
+			if elem.Name == got.Name {
+				found = true
+			}
+		}
+
+		assert.Equalf(t, true, found, "output didn't match: dump %s is missing from the output", elem.Name)
```
Comment (paulcacheux) [inline pkg/security/security_profile/dump/manager_test.go:42]: same here

File: pkg/security/security_profile/dump/manager_test.go [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/DataDog/datadog-go/v5/statsd"
+
+	"github.com/DataDog/datadog-agent/pkg/security/config"
+)
+
+func compareListOfDumps(t *testing.T, out, expectedOut []*ActivityDump) {
```
Comment (paulcacheux) [inline pkg/security/security_profile/dump/manager_test.go:23]: I think `t.Helper()` would help pinpoint the actual error position

File: pkg/security/security_profile/dump/profile.go [hunk 14]
```diff
@@ -206,8 +206,8 @@ func (ad *ActivityDump) GenerateProfileData() Profile {
 	return p
 }
 
-// EncodeProfile encodes an activity dump in the Profile format
-func (ad *ActivityDump) EncodeProfile() (*bytes.Buffer, error) {
+// EncodeSecL encodes an activity dump in the Profile format
```
Comment (paulcacheux) [inline pkg/security/security_profile/dump/profile.go:209]:
```suggestion
// EncodeSecL encodes an activity dump in the Secl format
```
?

**Review by paulcacheux — APPROVED** [review]
Few nits, but otherwise looks good

File: pkg/security/config/dump.go (hunk 8 above)
Comment (Gui774ume) [reply pkg/security/config/dump.go:120]: Haha I actually prefer lower only, upper feels weird to me 😏

File: pkg/security/security_profile/dump/profile.go (hunk 14 above)
Comment (Gui774ume) [reply pkg/security/security_profile/dump/profile.go:209]: 👍🏻

File: pkg/security/security_profile/dump/manager_test.go (hunk 12 above)
Comment (Gui774ume) [reply pkg/security/security_profile/dump/manager_test.go:42]: 👍🏻

File: pkg/security/security_profile/dump/activity_dump.go (hunk 9 above)
Comment (Gui774ume) [reply pkg/security/security_profile/dump/activity_dump.go:298]: I think it does what's it's supposed to ... I can move things around but it will eventually do the same thing. Adding unit tests !

Comment (paulcacheux) [reply pkg/security/security_profile/dump/activity_dump.go:298]: I mean, what does `nodes = append(nodes[1:])` mean ? Isn't the same as `nodes = nodes[1:]`?

File: pkg/security/security_profile/dump/manager_test.go (hunk 13 above)
Comment (Gui774ume) [reply pkg/security/security_profile/dump/manager_test.go:23]: I actually just moved this file but somehow Git didn't pick up on the move for this one. Let keep the changes to a minimum ? The PR is already quite big :( But happy to make another PR with your fix if you want me to !

File: pkg/config/system_probe.go (hunk 7 above)
Comment (Gui774ume) [reply pkg/config/system_probe.go:302]: Happy to change to whatever you think might be better ! (you meant security snapshots no ?)

File: pkg/security/security_profile/dump/activity_dump.go (hunk 9 above)
Comment (Gui774ume) [reply pkg/security/security_profile/dump/activity_dump.go:298]: I changed the algo a little bit to make it clearer 😃

**Review by spikat — APPROVED** [review]

File: pkg/security/security_profile/profile/manager.go [hunk 15]
```diff
@@ -0,0 +1,368 @@
...
+	// instantiate directory provider
+	if len(config.SecurityProfileDir) != 0 {
+		dirProvider, err := NewDirectoryProvider(config.SecurityProfileDir, config.SecurityProfileWatchDir)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't instantiate a new security profile directory provider: %w", err)
+		}
+		providers = append(providers, dirProvider)
+	}
+
+	profileCache, err := simplelru.NewLRU[cgroupModel.WorkloadSelector, *SecurityProfile](config.SBOMResolverWorkloadsCacheSize, nil)
```
Comment (safchain) [inline pkg/security/security_profile/profile/manager.go:57]: why using `config.SBOMResolverWorkloadsCacheSize`

File: pkg/security/security_profile/profile/manager.go [hunk 16]
```diff
@@ -0,0 +1,368 @@
...
+		}
+	}
+
+	// register the manager to the CGroup resolver
+	_ = m.cgroupResolver.RegisterListener(cgroup.WorkloadSelectorResolved, m.OnWorkloadSelectorResolvedEvent)
+	_ = m.cgroupResolver.RegisterListener(cgroup.CGroupDeleted, m.OnCGroupDeletedEvent)
+
+	seclog.Infof("security profile manager started")
+
+	for {
```
Comment (safchain) [inline pkg/security/security_profile/profile/manager.go:96]: you can remove the for loop I think

Comment (Gui774ume) [reply pkg/security/security_profile/profile/manager.go:96]: nice catch ... bad copy paste I guess ! 🙇🏻

File: pkg/security/security_profile/profile/manager.go (hunk 15 above)
Comment (Gui774ume) [reply pkg/security/security_profile/profile/manager.go:57]: Because it's a mistake 🤦🏻 Nice catch ! 👍🏻

**Review by safchain — APPROVED** [review]

File: pkg/security/security_profile/profile/manager.go [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+}
+
+// OnWorkloadSelectorResolvedEvent is used to handle the creation of a new cgroup with its resolved tags
+func (m *SecurityProfileManager) OnWorkloadSelectorResolvedEvent(workload *cgroupModel.CacheEntry) {
+	m.profilesLock.Lock()
+	defer m.profilesLock.Unlock()
+	workload.Lock()
+	defer workload.Unlock()
+
+	if workload.Deleted.Load() {
```
[the remark this answers is not in the document]
Comment (Gui774ume) [reply pkg/security/security_profile/profile/manager.go:123]: (for future reference) => we decided to delete the workload.Deleted check in `CheckTags` in the cgroup resolver as we don't want to mask an event that listeners might be dependent on to free async / cached resources.

**Review by amenasria — APPROVED** [review]
LGTM for Agent Platform owned file !
